Disable escape in firefox player

smason opened the pull request:
Currently this just watches for Firefox's pop-out media player and "disables" the Escape key when it's active.

This is useful because I use a vi like text editor (Helix) and I have a habit of hitting Esc everywhere.  This causes the video player to close which isn't what I want to do (I can always use my real keybinding for closing a window if I really want to close it!).

smason commented:
I use PEP-636 style Structural Pattern Matching which means it relies on having at least Python 3.10.  Would be easy to relax this if somebody else wanted to run with something older.

OctopusET commented:
Thank you for your work
Would you squash your commits into one?

smason committed:
Python script to selectively disable keys according to focus

Currently this just watches for Firefox's pop-out media player and
disables the Escape key when it's active.

This is useful because I use a vi like text editor (Helix) and I have a
habit of hitting Esc everywhere.  This causes the video player to close
which isn't what I want to do (I can always use my real keybinding for
closing a window if I really want to close it!).

smason commented:
> Would you squash your commits into one?

Should be done, wasn't sure if there was other changes do do first so left it kind of bare.

OctopusET commented on firefox-focus-monitor.py:
```diff
@@ -0,0 +1,94 @@
+    "determine whether we should bind Escape key"
+    logging.debug("received event %s", event.ipc_data)
+    match event.container:
+        case i3ipc.Con(app_id="firefox", name="Picture-in-Picture"):
```
This line will not work for some other languages such as Korean.
The name in Korean is "화면 속 화면".

But it seems like there's no other solution to determine if it's a PIP window or not other than this one. We need to add some comment for non-English users.

OctopusET replied:
Or you can actually expend features of this script to ignore keys for certain windows/programs

smason replied:
I spent a while looking for ways to get control over this before ending up in Python...  https://superuser.com/q/1812355/171763 was hoping somebody else had ideas.

Yup, matching on human readable names doesn't seem great but it worked for me in English!  Presume this is what X11 window class names were designed for, but don't remember them being very useful in practice.

A config file might make sense, but might be worth getting some more use cases before designing anything too fancy.

OctopusET commented on firefox-focus-monitor.py:
```diff
@@ -0,0 +1,94 @@
+import argparse
```
> I use PEP-636 style Structural Pattern Matching which means it relies on having at least Python 3.10.

Let's add some comments about the minimum requirements for the Python3 version.

If there's a lot of python scripts in this repository later. I will add some minimum requirements around the whole project.

smason replied:
just thought that a lot of the value from "contrib" style repos is that it's easy to grab odd bits of the code out for inspiration for your own use case.  having something too polished and packaged up seems to go against that...

OctopusET replied:
I agree, but I also don't want this repository to have too many broken things.

I won't impose strict rules, but let's ensure that people know how to use it and establish minimum requirements for this script.

The minimum requirement for this repository is just an idea. I guess it would be better not to have it.

So, I suggest you add some comments regarding the Python version requirements, like:
` python --version >= Python 3.10.0`
Also, please copy the comments from your commit and paste them at the top of your script.

smason replied:
Would be easy to set up some CI actions for automatic linting of code.  https://github.com/astral-sh/ruff seems to be the fashionable option for Python at the moment and I can confirm it works well with Github actions (and is easy to run locally as well)

OctopusET replied:
Oh thanks, that's what I've been looking for.

smason committed:
add module level description

smason committed:
use correct logger

OctopusET commented:
I will check this PR and other your PR on this weekend. Thanks for your PR.
I think [af1e206](https://github.com/OctopusET/sway-contrib/pull/13/commits/af1e2065cb79e32f97324f8dd5541aa7d25e6f56) should be squashed to previous commit or split the PR first. First PR for disabling escape in firefox player and second PR for the lint.

And also I'm considering using merge PR with squash. Personally I don't prefer merge commit. This project doesn't have to be strict on commit history, but still I don't want it make it mess. Please share your idea people. I'd like to see how other projects are managing PR like this one.

smason committed:
Make it easier to add other binds

Not sure who'd actually want to use it, but it's enough to add calls
like:

   mon.bind("Space", app_id="foot")

to prevent spacebar from being passed through to the foot terminal
emulator.

smason commented:
> I will check this PR and other your PR on this weekend. Thanks for your PR. I think [af1e206](https://github.com/OctopusET/sway-contrib/pull/13/commits/af1e2065cb79e32f97324f8dd5541aa7d25e6f56) should be squashed to previous commit or split the PR first. 

yup, squashing makes sense.  have squashed the top two commits

> And also I'm considering using merge PR with squash. Personally I don't prefer merge commit. This project doesn't have to be strict on commit history, but still I don't want it make it mess. Please share your idea people. I'd like to see how other projects are managing PR like this one.

The upstream [sway contributor guidelines](https://github.com/swaywm/sway/blob/master/CONTRIBUTING.md) don't say much about when to squash.  Just had a look at my goto exemplar of a well run project, [trio](https://github.com/python-trio/trio), and their [contributing guidelines](https://trio.readthedocs.io/en/latest/contributing.html) but it doesn't say much either.  They're both much bigger projects so presume it's not much of a concern even of project's their size.

I personally don't care too much about commit contents (as evidenced by the number of mine) so long as they don't get too big and aren't combining unrelated changes.  If this was a much bigger project with hundreds or thousands of contributors then things would be different, but given it's only a few people making relatively infrequent contributions I, personally, wouldn't worry too much about it.

OctopusET reviewed (approved):
LGTM
Sorry for the delays
